File: clients/macos/Tests/ChatOSCoreTests/QuickSearchRankingTests.swift
import ChatOSCore
import Testing

struct QuickSearchRankingTests {
    @Test
    func exactAndPrefixMatchesOutrankFuzzyMatches() throws {
        let exact = try #require(QuickSearchRanking.score(query: "Safari", title: "Safari"))
        let prefix = try #require(QuickSearchRanking.score(query: "Saf", title: "Safari"))
        let fuzzy = try #require(QuickSearchRanking.score(query: "sfr", title: "Safari"))

        #expect(exact > prefix)
        #expect(prefix > fuzzy)
    }

    @Test
    func chineseSubstringMatchesWithoutTokenization() throws {
        let score = try #require(QuickSearchRanking.score(
            query: "项目",
            title: "打开项目设置"
        ))
        #expect(score > 200)
    }

    @Test
    func unrelatedTextDoesNotMatch() {
        #expect(QuickSearchRanking.score(query: "terminal", title: "Safari") == nil)
    }

    @Test
    func recencyCannotOvertakeAHighlyAccurateMatch() throws {
        let exact = try #require(QuickSearchRanking.score(query: "Notes", title: "Notes"))
        let fuzzyRecent = try #require(QuickSearchRanking.score(
            query: "nts",
            title: "Network Tools",
            recencyBoost: 1_000,
            frequencyBoost: 1_000
        ))
        #expect(exact > fuzzyRecent)
    }
}

File: clients/windows/tests/ChatOS.Core.Tests/QuickSearchRankingTests.cs
using ChatOS.Core.Domain;

namespace ChatOS.Core.Tests;

public sealed class QuickSearchRankingTests
{
    [Fact]
    public void ExactAndPrefixMatchesOutrankFuzzyMatches()
    {
        var exact = QuickSearchRanking.Score("Edge", "Edge");
        var prefix = QuickSearchRanking.Score("Ed", "Edge");
        var fuzzy = QuickSearchRanking.Score("eg", "Edge");

        Assert.NotNull(exact);
        Assert.NotNull(prefix);
        Assert.NotNull(fuzzy);
        Assert.True(exact > prefix);
        Assert.True(prefix > fuzzy);
    }

    [Fact]
    public void ChineseSubstringMatchesWithoutTokenization() =>
        Assert.True(QuickSearchRanking.Score("项目", "打开项目设置") > 200);

    [Fact]
    public void UnrelatedTextDoesNotMatch() =>
        Assert.Null(QuickSearchRanking.Score("terminal", "Edge"));

    [Fact]
    public void BoostsCannotOvertakeHighlyAccurateMatch()
    {
        var exact = QuickSearchRanking.Score("Notes", "Notes");
        var fuzzy = QuickSearchRanking.Score("nts", "Network Tools", recencyBoost: 1_000, frequencyBoost: 1_000);
        Assert.True(exact > fuzzy);
    }
}
